tests/test_mnemonic_generation.py

```diff
@@ -195,3 +218,95 @@ def test_50_dice_rolls():
+    coin_flips = "10101010101011101100010000000011001000110000000000110000000010011100000000000000100000000001100000011100101100100001001100110010"
+    expected = "primary involve absorb ecology adapt agent abandon avoid blossom tortoise luggage grape"
+    mnemonic = mnemonic_generation.generate_mnemonic_from_coin_flips(coin_flips)
+    actual = " ".join(mnemonic)
+    assert bip39.mnemonic_is_valid(actual)
+    assert actual == expected
+
+    coin_flips = "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
+    expected = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about"
+    mnemonic = mnemonic_generation.generate_mnemonic_from_coin_flips(coin_flips)
+    actual = " ".join(mnemonic)
+    assert bip39.mnemonic_is_valid(actual)
+    assert actual == expected
+
+    coin_flips = "10100010100101010110101011010001010011101011111010010101010010101010101010010100101010101001010101010110101011111111000001001000"
+    expected = "penalty prize reform output true pipe prevent next next rely winter museum"
+    mnemonic = mnemonic_generation.generate_mnemonic_from_coin_flips(coin_flips)
+    actual = " ".join(mnemonic)
+    assert bip39.mnemonic_is_valid(actual)
+    assert actual == expected
+
+    
+def test_coin_flips_invalid_input():
+    """
+        Should raise a ValueError if coin_flips contains invalid characters or has an invalid length.
+    """
+    # Test invalid characters
+    with pytest.raises(ValueError) as e:
+        invalid_coin_flips = "01012" + "0" * 123  # 128 chars with a '2'
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "only contain '0' or '1'" in str(e)
+
+    with pytest.raises(ValueError) as e:
+        invalid_coin_flips = "01ab" + "0" * 252  # 256 chars with 'a' and 'b'
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "only contain '0' or '1'" in str(e)
+
+    # Test invalid lengths
+    with pytest.raises(ValueError) as e:
+        invalid_coin_flips = "0101"  # Too short (4 chars)
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "128 or 256 bits" in str(e)
+
+    with pytest.raises(ValueError) as e:
+        invalid_coin_flips = "0" * 127  # Too short (127 chars)
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "128 or 256 bits" in str(e)
+
+    with pytest.raises(ValueError) as e:
+        invalid_coin_flips = "0" * 129  # Too long for 128, too short for 256
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "128 or 256 bits" in str(e)
+
+    with pytest.raises(ValueError) as e:
+        invalid_coin_flips = "0" * 255  # Too short for 256
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "128 or 256 bits" in str(e)
```

Comment:
I would refactor all of these into another list of tuples/dicts to streamline the test code.

Somewhere around 3 repetitions should be the point where you consider refactoring to be more DRY.